lta/nersc_verifier.py

```diff
@@ -0,0 +1,240 @@
+            # note: 'lta' is an application-private metadata field
+            "lta": bundle,
+        }
+        # add the bundle file to the File Catalog
+        self.logger.info(f"POST /api/files - {hpss_path}")
+        await fc_rc.request("POST", "/api/files", file_record)
+        # for each file contained in the bundle
+        for fc_file in bundle["files"]:
+            fc_file_uuid = fc_file["uuid"]
+            # read the current file entry in the File Catalog
+            fc_record = await fc_rc.request("GET", f"/api/files/{fc_file_uuid}")
+            # add a location indicating the bundle archive
+            locations = fc_record["locations"]
+            fc_filename = os.path.basename(fc_record["logical_name"])
+            locations.append({
+                "site": "NERSC",
+                "path": f"{hpss_path}:{fc_filename}",
+                "archive": True,
+            })
+            # save the updated entry back to the File Catalog
+            await fc_rc.request("PATCH", f"/api/files/{fc_file_uuid}", locations)
+        # indicate that our file catalog updates were successful
+        return True
+
+    async def _update_bundle_in_lta_db(self, lta_rc: RestClient, bundle: BundleType) -> bool:
+        """Update the LTA DB to indicate the Bundle is verified."""
+        bundle_id = bundle["uuid"]
+        bundle["status"] = "completed"
+        bundle["update_timestamp"] = now()
+        bundle["claimed"] = False
+        self.logger.info(f"PATCH /Bundles/{bundle_id} - '{bundle}'")
+        await lta_rc.request('PATCH', f'/Bundles/{bundle_id}', bundle)
+        # the morning sun has vanquished the horrible night
+        return True
+
+    async def _verify_bundle_in_hpss(self, lta_rc: RestClient, bundle: BundleType) -> bool:
+        """Verify the checksum of the bundle in HPSS."""
+        bundle_id = bundle["uuid"]
+        # determine the path where it is stored on hpss
+        data_warehouse_path = bundle["path"]
+        basename = os.path.basename(bundle["bundle_path"])
+        stupid_python_path = os.path.sep.join([self.tape_bath_path, data_warehouse_path, basename])
+        hpss_path = os.path.normpath(stupid_python_path)
+        # run an hsi command to calculate the checksum of the archive as stored
+        #     hashverify    -> Verify checksum hash for existing HPSS file(s)
+        #     -A            -> enable auto-scheduling of retrievals
+        #     -C            -> cache purge option. If specified, purge files
+        #                      from HPSS disk cache after a successful hash
+        #                      create. Normally used only when it's expected
+        #                      that files will only be fetched once (or infrequently),
+        #                      to help optimize disk cache use
+        #     -H sha512     -> specify that the SHA512 algorithm be used to calculate the checksum
+        args = ["hsi", "hashverify", "-A", "-C", "-H", "sha512", hpss_path]
```

Comment:
:warning: We may want to ask NERSC about this.  If the system is reading the entire tape into cache at once, and we're continually erasing the cache, this could result in poor performance.

Reply:
Yeah, you're probably right on this. I thought `-C` would be a nice hint to HPSS, like "we're going to calculate the hash, but not otherwise use this file". However, the LTA component has a very narrow scope, and so has limited information to make global decisions like cache eviction. I'll remove it.